### PR TITLE
try to solve #13381

### LINKED ISSUE #13381: Reason for size limit on imread
opencv 4.0
windows 10 64 bits visual 2017

[My test program](http://answers.opencv.org/question/203793) : 
```
    RNG r;
    Mat m(60000, 40000, CV_8UC3);
    r.fill(m, RNG::UNIFORM, 0, 255);
    imwrite("test.png", m);
    Mat p = imread("test.png", IMREAD_UNCHANGED),d;
    cout << p.rows << " " << p.cols;
    absdiff(m, p, d);
    Scalar v = mean(d);
    cout << " mean=" << v << "\n";

```
and an exception is thrown (opencv 4.0.0 windows 10 visual 2017 win64) Now I changed in source code

#define CV_IO_MAX_IMAGE_PIXELS (1<<30) // 1 Gigapixel

in

#define CV_IO_MAX_IMAGE_PIXELS (1ull<<32ull) // 1 Gigapixel

and no exception is thrown (after opencv rebuild)

### PR DESCRIPTION
**Pull request by LaurentBerger**
resolves #13381

### CODE REVIEW

File: modules/imgcodecs/src/loadsave.cpp [hunk 1]
```diff
@@ -60,18 +61,16 @@ namespace cv {
 
 // TODO Add runtime configuration
 #define CV_IO_MAX_IMAGE_PARAMS (50)
-#define CV_IO_MAX_IMAGE_WIDTH (1<<20)
-#define CV_IO_MAX_IMAGE_HEIGHT (1<<20)
```
Comment (alalek) [inline modules/imgcodecs/src/loadsave.cpp]: Please remove "TODO" comment too and perform reading of these options once ([example](https://github.com/opencv/opencv/blob/3.4.4/modules/core/src/ocl.cpp#L223-L226), but probably inside of `validateInputImageSize()` function).

File: modules/imgcodecs/src/loadsave.cpp [hunk 2]
```diff
@@ -60,18 +61,16 @@ namespace cv {
...
     CV_Assert(size.width > 0);
-    CV_Assert(size.width <= CV_IO_MAX_IMAGE_WIDTH);
+    CV_Assert(static_cast<size_t>(size.width) <= utils::getConfigurationParameterSizeT("OPENCV_IO_MAX_IMAGE_WIDTH", 1<<20));
     CV_Assert(size.height > 0);
-    CV_Assert(size.height <= CV_IO_MAX_IMAGE_HEIGHT);
+    CV_Assert(static_cast<size_t>(size.height) <= utils::getConfigurationParameterSizeT("OPENCV_IO_MAX_IMAGE_HEIGHT", 1<<20));
     uint64 pixels = (uint64)size.width * (uint64)size.height;
-    CV_Assert(pixels <= CV_IO_MAX_IMAGE_PIXELS);
+    uint64 minSizeFile = utils::getConfigurationParameterSizeT("OPENCV_IO_MAX_IMAGE_WIDTH", 1 << 20)*utils::getConfigurationParameterSizeT("OPENCV_IO_MAX_IMAGE_HEIGHT", 1 << 20);
+    CV_Assert(pixels <= utils::getConfigurationParameterSizeT("OPENCV_IO_MAX_IMAGE_PIXELS", std::min<size_t>(minSizeFile, std::numeric_limits<std::size_t>::max()>>2)));
```
Comment (alalek) [inline modules/imgcodecs/src/loadsave.cpp]: 1. `minSizeFile` is not needed. Max width, max height and max total number of pixels should stay different. Nobody want to allocate 128Gb on server due malformed input image.

2. `std::numeric_limits<std::size_t>::max()>>2` - this threshold is very huge, so please keep defaults the same.

Comment (LaurentBerger) [reply modules/imgcodecs/src/loadsave.cpp]: OK

**Comment by LaurentBerger** [comment]
Finally if I don't change  CV_IO_MAX_IMAGE_PIXELS (1<<30) exception still occurs then can we change in 1<<31 ?

**Comment by alalek** [comment]
@LaurentBerger Right. This behavior is expected. There is another description of the problem: #8511

**Comment by LaurentBerger** [comment]
Sorry my idea is bad. I don't change default value and I add a note in imread doc : 

> By default number of pixels must be less than 2^30. Limit can be set using system variable OPENCV_IO_MAX_IMAGE_PIXELS

**Comment by alalek** [comment]
[Documentation note](http://pullrequest.opencv.org/buildbot/export/pr/13382/docs/d4/da8/group__imgcodecs.html#ga288b8b3da0892bd651fce07b3bbd3a56) :+1:

**Pull request merged**